cli/klaus/src/main.rs

```diff
@@ -94,31 +126,61 @@ async fn main() -> Result<(), Error> {
     }
     register_activation(&args.registry_path, &args.dot_flox_hash, activation)?;
 
-    // Start the listeners
-    let shutdown_flag = Arc::new(AtomicBool::new(false));
-    let signal_listener = signal_listener()?;
-    let signal_task = spawn_signal_listener(signal_listener, shutdown_flag.clone())?;
-    let termination_task = spawn_termination_listener(activation.into(), shutdown_flag.clone());
-
     info!(
         this_pid = nix::unistd::getpid().as_raw(),
         target_pid = args.pid,
         "watchdog is on duty"
     );
 
     // Listen for a notification
-    let action = listen(signal_task, termination_task, shutdown_flag).await;
-
-    match action {
-        Action::Cleanup => {
-            info!(pid = &args.pid, "deregistering activation");
-            deregister_activation(&args.registry_path, &args.dot_flox_hash, activation)
-                .context("failed to deregister activation")?;
-        },
-        Action::Terminate => {
-            // TODO: deregister if !is_running?
-            debug!("received termination action, exiting");
-        },
+    #[cfg(target_os = "macos")]
+    let res = wait_for_termination(watcher, should_proceed, should_stop);
+
+    #[cfg(target_os = "linux")]
+    let res = wait_for_termination(should_proceed, should_stop);
+
+    if res.is_err() {
+        error!("received stop signal, exiting");
+        return res;
+    }
+
+    // Now we proceed assuming we've gotten a termination notification.
+    // If we get a SIGINT/SIGTERM/SIGQUIT/SIGKILL we leave behind the activation in the registry,
+    // but there's not much we can do about that because we don't know who sent us one of those
+    // signals or why.
+
+    let remaining_activations =
+        deregister_activation(&args.registry_path, &args.dot_flox_hash, activation)
+            .context("failed to deregister activation")?;
+    debug!(n = remaining_activations, "remaining activations");
+    if remaining_activations == 0 {
+        if args.socket_path.exists() {
+            let mut cmd = Command::new(&*PROCESS_COMPOSE_BIN);
```

Comment:
(non-blocking) I've made a note in https://github.com/flox/flox/issues/1594 that we might want to expose this as a function in the provider.

Reply:
For a different day